synapse/util/logcontext.py

```diff
@@ -311,6 +311,9 @@ def preserve_context_over_deferred(deferred, context=None):
     """Given a deferred wrap it such that any callbacks added later to it will
     be invoked with the current context.
     """
+    if not isinstance(deferred, defer.Deferred):
```

Comment:
I am not in favour of further bodges on top of `preserve_context_over_deferred`, since afaict it is broken. Let me send you a counter-PR here.